### PR TITLE
WIP: Serialize structs

### PR DESCRIPTION
**Pull request by antoinealb**
This pull request is a work in progress but I started working on header file generation.

I am not sure about the best way to implement tagged unions. What do @pavel-kirienko think ?

### CODE REVIEW

**Comment by pavel-kirienko** [comment]
This is a good start.

### Template engine

I suggest to avoid external dependencies as much as we can, because we don't want to ask users to install additional stuff just to get their application to compile. [Libuavcan solves this by employing Pyratemp](https://github.com/UAVCAN/libuavcan/tree/master/libuavcan/dsdl_compiler/libuavcan_dsdl_compiler) - a MIT-licensed single-file Python template engine. In terms of template syntax, Pyratemp is a very special snowflake, so I wrote an adapter that pre-processes the template code and converts it from Mako-style syntax (which personally I find more human-friendly than Jinja, but unto each their own) into Pyratemp syntax. The adapter can be found here: https://github.com/UAVCAN/libuavcan/blob/master/libuavcan/dsdl_compiler/libuavcan_dsdl_compiler/__init__.py#L245-L309.

I'm not sure if we need to use this adapter here, but regardless, I strongly suggest to ditch Jinja and stick to dependency-free solutions.

### Constants

I see that constant generation is not implemented yet. Do you have a naming convention in mind? I would propose something like: `<full type name>_<field name>`.

For example, consider `uavcan.protocol.NodeStatus.HEALTH_OK`:

```
#define uavcan_protocol_NodeStatus_HEALTH_OK 0
```

Yes I too think that this is a bit ugly, but hey it's C, what would you expect.

### Function names

I noticed that serialization functions are named in the form `[de]serialize_<type name>`. Maybe we should consider swapping the parts into `<type name>__[de]serialize` (notice two underscores), because then all entities related to the same type will be prefixed identically. Namely:

- Serialization/deserialization functions
- Default data type ID
- Full type name
- Data type signature

The only downside is that we'll have to use double underscore in order to avoid name conflicts with type fields.

### Unions

My proposition would be quite straightforward:

```c
typedef struct
{
    union
    {
        FieldType1 foo;
        FieldType2 bar;
    };
    /*
     * The tag should be placed after the union for alignment reasons.
     * Leading underscore is needed to avoid name conflicts.
     */
    uint16_t _tag;
} MyUnion;

#define MyUnion__tag_foo 0
#define MyUnion__tag_bar 1
```

File: header.jinja [hunk 1]
```diff
@@ -0,0 +1,23 @@
+#ifndef _CANARD_SERIALIZE_
+#define _CANARD_SERIALIZE_
+
+#include <stdint.h>
+#include <stdbool.h>
+
+{% for t in types %}
+{{ uavcan_type_to_c(t) }} {
```
Comment (pavel-kirienko) [inline header.jinja]: This probably supposed to be like:

```c
typedef struct
{
    ...
} {{ uavcan_type_to_c(t) }};
```

**Comment by pavel-kirienko** [comment]
Is my understanding correct that you want to put all definitions into one huge header file, rather than scattering them into multiple header files like libuavcan does?

**Comment by pavel-kirienko** [comment]
Hey @antoinealb, this was a great start, but do you plan on continuing it though? No rush, just asking. ;)

**Comment by paulpatience** [comment]
One problem with the double underscore is that identifiers containing a double underscore are always reserved in C++.

**Pull request closed without merging**